### PR TITLE
Report a metric for time taken to perform recursive permission change

### PR DESCRIPTION
**Pull request by JornShen**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

ref: https://github.com/kubernetes/kubernetes/issues/95588

**Special notes for your reviewer**:
/cc @gnufied @jsafrane @mattcary @msau42

**Does this PR introduce a user-facing change?**:

```release-note
add a metric for time taken to perform recursive permission change
```

### CODE REVIEW

File: pkg/volume/awsebs/aws_ebs.go [hunk 1]
```diff
@@ -428,7 +428,7 @@ func (b *awsElasticBlockStoreMounter) SetUpAt(dir string, mounterArgs volume.Mou
 	}
 
 	if !b.readOnly {
-		volume.SetVolumeOwnership(b, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy)
+		volume.SetVolumeOwnership(b, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.OperationCompleteHook(b.plugin.GetPluginName(), "volume_fsgroup_recursive_apply"))
```
Comment (jsafrane) [inline pkg/volume/awsebs/aws_ebs.go:431]: Please use a const for "volume_fsgroup_recursive_apply". Or add `util.FSGroupCompleteHook()` into volume_linux.go 
 that would return `OperationCompleteHook` with the right op. name.

Comment (JornShen) [reply pkg/volume/awsebs/aws_ebs.go]: done.

File: pkg/volume/csi/csi_mounter.go [hunk 2]
```diff
@@ -278,7 +279,7 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	}
 
 	if c.supportsFSGroup(fsType, mounterArgs.FsGroup, c.fsGroupPolicy) {
-		err := volume.SetVolumeOwnership(c, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy)
+		err := volume.SetVolumeOwnership(c, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.FSGroupCompleteHook(c.plugin.GetPluginName()))
```
Comment (msau42) [inline pkg/volume/csi/csi_mounter.go:282]: should the plugin name be "csi", or should it be the name of the csi driver?

File: pkg/volume/flexvolume/mounter.go [hunk 3]
```diff
@@ -93,7 +94,7 @@ func (f *flexVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterArgs)
 
 	if !f.readOnly {
 		if f.plugin.capabilities.FSGroup {
-			volume.SetVolumeOwnership(f, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy)
+			volume.SetVolumeOwnership(f, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.FSGroupCompleteHook(f.plugin.GetPluginName()))
```
Comment (msau42) [inline pkg/volume/flexvolume/mounter.go:97]: Should the plugin name be "flexvolume" or the name of the flex driver?

**Review by msau42 — COMMENTED** [review]
Thanks for working on the change so quickly! I have some followup questions and comments. Should we have a unit or e2e test to make sure we're properly setting these metrics. I believe we have some unit and/or e2e tests for other metrics already. @gnufied could you help point to the existing tests?

File: pkg/volume/csi/csi_mounter.go (hunk 2 above)
Comment (mattcary) [reply pkg/volume/csi/csi_mounter.go:282]: +1, it should probably be c.driverName or maybe pluginName/driverName

Comment (JornShen) [reply pkg/volume/csi/csi_mounter.go:282]: ohh, sorry! I will take other PR to fix it.  Thanks for pointing at it!

Comment (gnufied) [reply pkg/volume/csi/csi_mounter.go:282]: +1 to capturing full plugin name. we should use `GetFullQualifiedPluginNameForVolume` I think.

File: pkg/volume/flexvolume/mounter.go (hunk 3 above)
Comment (JornShen) [reply pkg/volume/flexvolume/mounter.go:97]: I change it to "kubernetes.io/flexvolume:[flex driver name]". see https://github.com/kubernetes/kubernetes/pull/95892/.

File: pkg/volume/csi/csi_mounter.go (hunk 2 above)
Comment (JornShen) [reply pkg/volume/csi/csi_mounter.go:282]: IMO,  c.driverName is more detailed maybe.

https://github.com/kubernetes/kubernetes/blob/87cc9bca0fcbbdf1b8232c9c71e50023d62667c3/pkg/volume/csi/csi_plugin.go#L357-L370 

c.driverName gets driver name from `CSIVolumeSource` and `CSIPersistentVolumeSource` 

While, `GetFullQualifiedPluginNameForVolume` get driver name from `CSIPersistentVolumeSource`

https://github.com/kubernetes/kubernetes/blob/87cc9bca0fcbbdf1b8232c9c71e50023d62667c3/pkg/volume/util/metrics.go#L122-L127

See https://github.com/kubernetes/kubernetes/pull/95892

**Pull request merged**